### PR TITLE
Setup basic frontend

### LINKED ISSUE #5: Setup basic frontend
- [x] Webpack
- [x] TailwindCSS

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/5

## What happened 👀

Setting up `Webpack` and `TailwindCSS` to the project.

## Insight 📝

- [x] Webpack
- [x] TailwindCSS

Extra:
- Upgrade node to the latest version (15.5.1), it's about the security
- Update Dockerfile to build asset with Webpack then merge to the main container

## Proof Of Work 📹

Running the website with docker.
Screen Shot 2564-01-06 at 15 18 24

### CODE REVIEW

File: assets/stylesheets/application.scss [hunk 1]
```diff
@@ -1,3 +1,7 @@
+@import "tailwindcss/base";
+@import "tailwindcss/components";
```
Comment (hoangmirs) [inline assets/stylesheets/application.scss]: Do you have any problems when running on development mode and try changing some code?
I had a problem when running with `--watch`, and it took 16 seconds to rebuild the assets

File: assets/stylesheets/application.scss [hunk 2]
```diff
@@ -1,3 +1,7 @@
...
+@import "tailwindcss/utilities";
+
 body {
```
Comment (hoangmirs) [inline assets/stylesheets/application.scss:5]: Should we move the base style into another file?

File: assets/stylesheets/application.scss (hunk 1 above)
Comment (Lahphim) [reply assets/stylesheets/application.scss]: @hoangmirs I do have those problems 🤣

File: assets/stylesheets/application.scss (hunk 2 above)
Comment (Lahphim) [reply assets/stylesheets/application.scss:5]: @hoangmirs I just did a simple things, I'll remove these once I'm working on feature tickets ✌️

File: conf/webpack/production.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      filename: 'stylesheets/[name].css',
+      chunkFilename: '[id].css'
+    })
+  ],
+  module: {
+    rules: [
+      {
+        test: /\.(sa|sc|c)ss$/,
+        use: [
+          MiniCssExtractPlugin.loader,
```
Comment (hoangmirs) [inline conf/webpack/production.js]: I faced an issue with this on production, so please take a look at [this PR](https://github.com/hoangmirs/go-scraper/pull/30) to avoid my issue

File: conf/webpack/common.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const CopyPlugin = require("copy-webpack-plugin");
+
+module.exports = {
+  entry: {
+    application: path.resolve(__dirname, '../..', 'assets', 'javascripts', 'application.js')
+  },
+  plugins: [
+    new CleanWebpackPlugin(),
+    new CopyPlugin({
```
Comment (hoangmirs) [inline conf/webpack/common.js]: Do we need this?
Before I used this, but I think we can add `/assets` folder as StaticPath, then we just use it like we did with Rails
https://beego.me/docs/mvc/view/static.md

File: go.mod [hunk 5]
```diff
@@ -5,6 +5,7 @@ go 1.15
 require github.com/beego/beego/v2 v2.0.1
 
 require (
+	github.com/ddollar/forego v0.16.1 // indirect
```
Comment (hoangmirs) [inline go.mod]: You added in Makefile `install-dependencies`, and this is not related to the code, so should we omit it?

File: README.md [hunk 6]
```diff
@@ -36,7 +37,16 @@ Visting http://localhost:8080/ with a web browser will display the application.
 $ make test
 ````
 
+## License
+This project is Copyright (c) 2014-2020 Nimble. It is free software,
+and may be redistributed under the terms specified in the [LICENSE] file.
+
+[LICENSE]: /LICENSE
+
+
```
Comment (junan) [inline README.md]: Could we remove extra space here?

Comment (Lahphim) [reply README.md]: @junan Fixed https://github.com/Lahphim/go-crawler-challenge/pull/33/commits/299435d346c848437cc801c9f272123f7e4941d8

File: go.mod (hunk 5 above)
Comment (Lahphim) [reply go.mod]: @hoangmirs Removed https://github.com/Lahphim/go-crawler-challenge/pull/33/commits/299435d346c848437cc801c9f272123f7e4941d8

File: conf/webpack/common.js (hunk 4 above)
Comment (Lahphim) [reply conf/webpack/common.js]: @hoangmirs Updated https://github.com/Lahphim/go-crawler-challenge/pull/33/commits/5915e9bf0f63201c51de40b56bfde5ec7b3e78a1

File: conf/webpack/production.js (hunk 3 above)
Comment (Lahphim) [reply conf/webpack/production.js]: @hoangmirs Changed https://github.com/Lahphim/go-crawler-challenge/pull/33/commits/0dbce57a6f067582175a967d85ccab61a6abc5f9

File: Dockerfile [hunk 7]
```diff
@@ -1,4 +1,15 @@
-FROM golang:1.15-alpine
+FROM node:15.5.1-alpine AS assets-builder
+
+WORKDIR /assets
+
+COPY ./package.json ./package-lock.json ./
+COPY ./conf/webpack/. ./conf/webpack/
+COPY ./assets/. ./assets/
```
Comment (hoangmirs) [inline Dockerfile]: You should copy tailwind and postCSS configurations as well, otherwise, it could not build the assets properly. Check it on [my PR](https://github.com/hoangmirs/go-scraper/pull/31)

File: README.md [hunk 8]
```diff
@@ -36,7 +37,15 @@ Visting http://localhost:8080/ with a web browser will display the application.
 $ make test
 ````
 
+## License
+This project is Copyright (c) 2014-2020 Nimble. It is free software,
```
Comment (hoangmirs) [inline README.md]: Let's forget 2020 😅

File: Dockerfile [hunk 9]
```diff
@@ -1,4 +1,13 @@
-FROM golang:1.15-alpine
+FROM node:15.5.1-alpine AS assets-builder
+
+WORKDIR /assets
+
+COPY . .
+
+RUN npm install && npm run build
+
```
Comment (hoangmirs) [inline Dockerfile:8]: Please remove this extra line

**Review by hoangmirs — APPROVED** [review]

**Review by junan — APPROVED** [review]
